diff -Naur miniupnpc-20080217/igd_desc_parse.c miniupnpcpatched/igd_desc_parse.c
--- miniupnpc-20080217/igd_desc_parse.c	2006-11-19 23:32:55.000000000 +0100
+++ miniupnpcpatched/igd_desc_parse.c	2008-02-17 13:17:43.000000000 +0100
@@ -17,6 +17,7 @@
 	struct IGDdatas * datas = (struct IGDdatas *)d;
 	memcpy( datas->cureltname, name, l);
 	datas->cureltname[l] = '\0';
+    //printf("element:%s\n",datas->cureltname);
 	datas->level++;
 }
 
@@ -60,6 +61,8 @@
            datas->level, datas->cureltname, l, data);	*/
 	if( !strcmp(datas->cureltname, "URLBase") )
 		dstmember = datas->urlbase;
+    if (!strcmp(datas->cureltname, "modelDescription") && datas->level <= 3)
+        dstmember = datas->modeldescription;
 	else if(datas->state<=1)
 	{
 		if( !strcmp(datas->cureltname, "serviceType") )
@@ -88,6 +91,9 @@
 	}
 	if(dstmember)
 	{
+//        printf("----->dstmember: %2d %s : %.*s\n",
+//           datas->level, datas->cureltname, l, data);	
+
 		if(l>=MINIUPNPC_URL_MAXSIZE)
 			l = MINIUPNPC_URL_MAXSIZE-1;
 		memcpy(dstmember, data, l);
diff -Naur miniupnpc-20080217/igd_desc_parse.h miniupnpcpatched/igd_desc_parse.h
--- miniupnpc-20080217/igd_desc_parse.h	2007-04-11 17:21:09.000000000 +0200
+++ miniupnpcpatched/igd_desc_parse.h	2008-02-17 13:17:43.000000000 +0100
@@ -27,6 +27,7 @@
 	char scpdurl[MINIUPNPC_URL_MAXSIZE];
 	char servicetype[MINIUPNPC_URL_MAXSIZE];
 	char devicetype[MINIUPNPC_URL_MAXSIZE];
+    char modeldescription[MINIUPNPC_URL_MAXSIZE];
 };
 
 void IGDstartelt(void *, const char *, int);
